### PR TITLE
MAINT: ``__array_interface__`` data address cannot be bytes

### PR DESCRIPTION
**Pull request by charris**
The old version 1 ``__array_interface__`` implementation allowed the
data pointer value to be passed as a string, in version 2 an integer was
required, but the old code accepting a string remained. Remove that bit
of code from PyArray_FromInterface.

### CODE REVIEW

**Review by eric-wieser — COMMENTED** [review]
Patch looks fine, I'm ok with this if we think we can get away with skipping a deprecation warning.

**Comment by charris** [comment]
@eric-wieser It's a bit of a gamble, but I think it is pretty safe. Folks making the transition to Python 3 would probably have noticed if they were using strings. I need to add a release note though.

File: doc/release/upcoming_changes/17241.compatibility.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+The ``__array_interface__`` data tuple first element must be an integer
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/17241.compatibility.rst:1]:
```suggestion
The first element of the ``__array_interface__["data"]`` tuple  must be an integer
```

**Comment by charris** [comment]
Updated.

**Comment by seberg** [comment]
I am OK with giving this a shot. Will probably merge this tomorrow, unless anyone has reason to be more cautious? Getting the correct string seems so tricky, that I am unsure where it would arise, besides maybe debugging where you copy paste a pointer representation?!

The main danger would be some library actually relying on it creating issues for a small group of third party users.

**Pull request merged**